attelo/decoding/mst.py

```diff
@@ -1,88 +1,116 @@
     scores = dict()
 
     for source, target, prob, rel in instances:
-        src = source.id
-        tgt = target.id
-        if tgt == root:
+        src, tgt = source.id, target.id
+
+        # Ignore all edges directed to the root
+        if tgt == root_id:
             continue
-        scores[src, tgt] = prob
+
+        scores[src, tgt] = min(MAX_SCORE, max(MIN_SCORE, logit(prob))) if use_prob else prob
         labels[src, tgt] = rel
-        if use_prob:  # probability scores
-            scores[src, tgt] = log(prob if prob != 0.0 else sys.float_info.min)
         targets[src].append(tgt)
 
     return Digraph(targets,
                    lambda s, t: scores[s, t],
-                   lambda s, t: labels[s, t]).mst()
-
+                   lambda s, t: labels[s, t])
 
-def _list_edges(instances, root='1', use_prob=True):
-    """ instances are quadruplets of the form:
+def _msdag(graph):
+    """ Returns a subgraph of graph (a Digraph) corresponding to its
+        Maximum Spanning Directed Acyclic Graph
 
-            source, target, probability_of_attachment, relation
-
-        root is the "root" of the graph, that is the node that has no incoming
-        nodes
-
-        returns a list of edges for the MST graph
+        Alogrithm is semi-greedy-MSDAG as described in Schluter (2014)
     """
-    mst = _graph(instances, root, use_prob=use_prob)
-
-    return [(src, tgt, mst.get_label(src, tgt))
-            for src, tgt in mst.iteredges()]
-
+    tree = graph.mst()
+    # Sort edges in orginal graph by decreasing score
+    edges = sorted(graph.iteredges(), key=lambda p:-graph.get_score(*p))
+    for src, tgt in edges:
+        # Already in graph ?
+        if tgt in tree.successors[src]:
+            continue
+        # Add the edge, revert if cycle is created
+        tree.successors[src].append(tgt)
+        if tree.find_cycle():
+            tree.successors[src].remove(tgt)
+
+    # Update score and label functions
+    # I really wish that Digraph used a pair as single argument,
+    #   the following would be clearer
+    nscores, nlabels = (dict((k, old(*k)) for k in tree.iteredges())
+                        for old in (graph.get_score, graph.get_label))
+
+    return Digraph(tree.successors,
+                   lambda s, t: nscores[s, t],
+                   lambda s, t: nlabels[s, t])
+
+def _list_edges(instances, use_prob=True, use_msdag=False):
+    """ Returns a list of edges for the MST/MSDAG graph """
+    
+    graph = _graph(instances, use_prob=use_prob)
+    subgraph = _msdag(graph) if use_msdag else graph.mst()
+
+    return [(src, tgt, subgraph.get_label(src, tgt))
+            for src, tgt in subgraph.iteredges()]
 
 class MstDecoder(Decoder):
     """ attach in such a way that the resulting subgraph is a
     maximum spanning tree of the original
     """
-    def __init__(self, root='1', use_prob=True):
-        self._root = root
+    def __init__(self, use_prob=True):
+        self._use_prob = use_prob
```

Comment:
There might be an issue here: now people are *forced* to take the textual first node as the root, which isn't always desirable.  In the RST DT case for example, we may want to take the fake root.

It's worth noting that the scikit version of attelo will accept a distinguished node name (probably `0` or maybe will change it to `ROOT` later) as corresponding to the fake root